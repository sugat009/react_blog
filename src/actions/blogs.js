import uuid from 'uuid';
import { database } from '../firebase/firebase';


// add blog post
export const addBlog = (blog) => ({
    type: 'ADD_BLOG',
    blog
});

// start add blog (async action for addBlog)
export const startAddBlog = (blogData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            title = '',
            content = '',
            createdAt = 0
        } = blogData;
        const blog = { title, content, createdAt };
        return database.ref(`users/${uid}/blogs`).push(blog)
            .then((ref) => {
                dispatch(addBlog({
                    id: ref.key,
                    ...blog
                }));
            });
    };
};

// remove blog post
export const removeBlog = ({ id } = {}) => ({
    type: 'REMOVE_BLOG',
    id
});

export const startRemoveBlog = ({ id } = {}) => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs/${id}`).remove().then(() => {
            dispatch(removeBlog({ id }));
        });
    };
};


// edit blog post
export const editBlog = (id, updates) => ({
    type: 'EDIT_BLOG',
    id,
    updates
});

export const startEditBlog = (id, updates) => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs/${id}`).update(updates)
            .then(() => {
                dispatch(editBlog(id, updates));
            });
    };
};


// SET Blogs part
export const setBlog = (blogs) => ({
    type: 'SET_BLOG',
    blogs
});

export const setStartBlog = () => {
    return (dispatch, getState) => {
        return database.ref(`users/${getState().auth.uid}/blogs`).once('value').then((snapshot) => {
            const blogData = [];
            snapshot.forEach((childSnapshot) => {
                blogData.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setBlog(blogData));
        });
    };  
};

// GET blogs part
export const getBlog = (blogs) => ({
    type: 'GET_BLOG',
    blogs
});

export const getStartBlog = () => {
    return (dispatch, getState) => {
        return database.ref(`users`).once('value').then((snapshot) => {
            const blogData = [];
            snapshot.forEach((childSnapshot) => {
                if (childSnapshot.key !== getState().auth.uid) {
                    blogData.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val().blogs
                    });
                }
            });
            dispatch(getBlog(blogData));
            console.log(blogData);
        });
    };
};