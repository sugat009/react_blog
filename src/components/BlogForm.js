import React from 'react';
import moment from 'moment';

export default class BlogForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            title: props.blog ? props.blog.title : '',
            content: props.blog ? props.blog.content : '',
            createdAt: props.blog ? moment(props.blog.createdAt) : moment(0),
            errors: ''
        };
    };
    contentChange = (e) => {
        const content = e.target.value;
        this.setState({
            content
        });
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState({
            title
        });
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.title || !this.state.content) {
            this.setState({
                errors: "Both title and content are required!"
            });
        } else {
            this.setState(() => ({ errors: ''}));
            if (moment(this.state.createdAt).isSame(moment(0))) {
                this.props.onSubmit({
                    title: this.state.title,
                    content: this.state.content,
                    createdAt: moment().valueOf()
                });
            } else {
                this.props.onSubmit({
                    title: this.state.title,
                    content: this.state.content
                });
            }
        }
    };
    render() {
        return (
            <form onSubmit={this.onSubmit} className="form">
                {this.state.errors && <p className="form__error">{this.state.errors}</p> }
                <input 
                    className="text-input"
                    type="text"
                    placeholder="Blog Title"
                    autoFocus
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <textarea 
                    className="text-area"
                    placeholder="Blog Content"
                    value={this.state.content}
                    onChange={this.contentChange}
                />
                <div>
                    <button className="button">Save Blog</button>
                </div>
            </form>
        )
    };
};