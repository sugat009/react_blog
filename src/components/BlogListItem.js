import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

const BlogListItem = ({ id, title, content, createdAt, user }) => (
    <Link to={`/edit/${id}`} className="list-item">
        <div>
            <h3 className="list-item__title">{title}</h3>
            <span className="list-item__subtitle">{moment(createdAt).format("MMMM Do, YYYY")}</span>
        </div>
        <h3 className="list-item__data">{user.displayName}</h3>
    </Link>
);

export default BlogListItem;