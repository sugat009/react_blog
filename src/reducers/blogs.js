const blogReducerDefaultState =  [];

export default (state = blogReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_BLOG':
            return [
                ...state,
                action.blog
            ];
        case 'REMOVE_BLOG':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_BLOG':
            return state.map((blog) => {
                if (blog.id === action.id) {
                    return {
                        ...blog,
                        ...action.updates
                    }
                } else {
                    return blog;
                };
            });
        case 'SET_BLOG':
            console.log(action.blogs, "aaaaaaaaaaaaaaaaaaaa");
            return action.blogs;
        case 'GET_BLOG':
            console.log(action.blogs);
            return action.blogs;
        default:
            return state;
    }
};