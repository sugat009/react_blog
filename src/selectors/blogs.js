import moment from 'moment';

// get visible blogs after the filters have been applied

export default (blogs, { title, sortBy, startDate, endDate }) => {
    return blogs.filter((blog) => {
        const createdAtMoment = moment(blog.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const titleMatch = blog.title.toLowerCase().includes(title.toLowerCase());
        
        return startDateMatch && endDateMatch && titleMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy === 'alphabet') {
            return a.title > b.title ? 1 : -1;
        }
    });
};