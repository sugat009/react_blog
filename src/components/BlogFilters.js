import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTitleFilter, sortByAlphabet, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class BlogFilters extends React.Component {
    state = {
        calenderFocused:  null
    };
    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'alphabetically') {
            this.props.sortByAlphabet();
        }
    };
    onTextChange = (e) => {
        this.props.setTitleFilter(e.target.value);
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }));
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            className="text-input"
                            type="text"
                            placeholder="search blogs"
                            value={this.props.filters.title}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select 
                            className="select"
                            value={this.props.filters.sortBy}
                            onChange={this.onSortChange}
                        >
                            <option value="alphabetically">Alphabetically</option>
                            <option value="date">Date</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker 
                            startDate = {this.props.filters.startDate}
                            endDate = {this.props.filters.endDate}
                            onDatesChange = {this.onDatesChange}
                            focusedInput = {this.state.calenderFocused}
                            onFocusChange = {this.onFocusChange}
                            numberOfMonths = {1}
                            isOutsideRange = {() => false}
                            showClearDates = {true}
                        />
                    </div>
                </div>
            </div>
        )
    };
};


const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTitleFilter: (title) => dispatch(setTitleFilter(title)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAlphabet: () => dispatch(sortByAlphabet()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(BlogFilters);