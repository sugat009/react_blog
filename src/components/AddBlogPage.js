import React from 'react';
import { connect } from 'react-redux';
import BlogForm from './BlogForm';
import { startAddBlog } from '../actions/blogs';


export class AddBlogPage extends React.Component {
    onSubmit = (blog) => {
        this.props.startAddBlog(blog);
        this.props.history.push("/");
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Blog</h1>
                    </div>
                </div>
                <div className="content-container">
                    <BlogForm onSubmit={this.onSubmit} />
                </div>
            </div>
        )
    };
};

const mapDispatchToProps = (dispatch) => ({
    startAddBlog: (blog) => dispatch(startAddBlog(blog))
});

export default connect(undefined, mapDispatchToProps)(AddBlogPage); // no mapStateToProps so undefined is used