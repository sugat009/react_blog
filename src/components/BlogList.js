import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';
import getVisibleExpenses from '../selectors/blogs';

export class BlogList extends React.Component {
    render() {
        var user = JSON.parse(localStorage.getItem('user'));
        return (
            <div className="content-container">
                <div className="list-header">
                    <div className="show-for-mobile">Blogs</div>
                    <div className="show-for-desktop">Blog</div>
                    <div className="show-for-desktop">Author</div>
                </div>
                <div className="list-body">
                    {
                        this.props.blogs.length === 0 ? (
                            <div className="list-item list-item--message">
                                <span>No Blogs Yet</span>
                            </div>
                        ) : (
                        this.props.blogs.map((blog) => {
                                return <BlogListItem key = {blog.id} {...blog} user={user} />
                        })
                        )
                    }
                </div>
            </div>
        );
    };
};

const mapStateToProps = (state) => ({
    blogs: getVisibleExpenses(state.blogs, state.filters)
});

export default connect(mapStateToProps)(BlogList);