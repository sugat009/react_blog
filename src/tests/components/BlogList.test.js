import React from 'react';
import { shallow } from 'enzyme';
import { BlogList } from '../../components/BlogList';
import blogs from '../fixtures/blogs';

test('should render BlogList with blogs', () => {
    const wrapper = shallow(<BlogList blogs={blogs} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render BlogList with no blogs', () => {
    const wrapper = shallow(<BlogList blogs={[]} />);
    expect(wrapper).toMatchSnapshot();
});