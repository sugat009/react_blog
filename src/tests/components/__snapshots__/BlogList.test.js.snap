// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BlogList with blogs 1`] = `
<div>
  <p>
    Blog List
  </p>
  <BlogListItem
    content="content of live your life"
    createdAt={432000000}
    id="1"
    key="1"
    title="live your life"
  />
  <BlogListItem
    content="fraand ill show u some ethical hacking"
    createdAt={-432000000}
    id="2"
    key="2"
    title="ethical hacking"
  />
  <BlogListItem
    content="Learn everything from scratch in C#"
    createdAt={0}
    id="3"
    key="3"
    title="C# MasterClass"
  />
</div>
`;

exports[`should render BlogList with no blogs 1`] = `
<div>
  <p>
    Blog List
  </p>
  <p>
    No Blogs Yet
  </p>
</div>
`;
